include/fuzzuf/optimizer/pso.hpp

```diff
@@ -0,0 +1,229 @@
+PSO<Dimension, ParticleNum>::PSO(double min_position, double max_position,
+                                 double min_velocity, double max_velocity,
+                                 double w, double c1, double c2)
+    : min_position(min_position),
+      max_position(max_position),
+      min_velocity(min_velocity),
+      max_velocity(max_velocity),
+      w(w),
+      c1(c1),
+      c2(c2) {}
+
+template <size_t Dimension, size_t ParticleNum>
+PSO<Dimension, ParticleNum>::~PSO() {}
+
+template <size_t Dimension, size_t ParticleNum>
+void PSO<Dimension, ParticleNum>::Init() {
+  for (auto& p : swarm) {
+    for (auto& pos : p.position)
+      pos = fuzzuf::utils::random::Random<double>(min_position, max_position);
+    p.velocity.fill(0);
+  }
+
+  idx = 0;
+  time = 0;
+}
+
+template <size_t Dimension, size_t ParticleNum>
+std::array<double, Dimension> PSO<Dimension, ParticleNum>::GetCurParticle() {
+  return swarm[idx].best_position;
+}
+
+template <size_t Dimension, size_t ParticleNum>
+void PSO<Dimension, ParticleNum>::SetScore(double score) {
+  auto& p = swarm[idx];
+  p.fitness = score;
+  UpdateLocalBest();
+
+  idx++;
+  idx %= swarm.size();
+
+  if (idx == 0) {
+    UpdateGlobalBest();
+    time++;
+  }
+}
+
+template <size_t Dimension, size_t ParticleNum>
+std::array<double, Dimension> PSO<Dimension, ParticleNum>::CalcValue() {
+  return best_position;
+}
+
+template <size_t Dimension, size_t ParticleNum>
+void PSO<Dimension, ParticleNum>::UpdatePositions() {
+  auto& p = swarm[idx];
+
+  for (size_t i = 0; i < Dimension; i++) {
+    double pos = p.position[i] + p.velocity[i];
+    pos = std::min(pos, max_position);
+    pos = std::max(pos, min_position);
+
+    p.position[i] = pos;
+  }
+}
+
+template <size_t Dimension, size_t ParticleNum>
+void PSO<Dimension, ParticleNum>::UpdateVelocities() {
+  auto& p = swarm[idx];
+
+  for (size_t i = 0; i < Dimension; i++) {
+    double v = w * p.velocity[i] +
+               c1 * fuzzuf::utils::random::Random<double>(0, 1) *
+                   (p.best_position[i] - p.position[i]) +
+               c2 * fuzzuf::utils::random::Random<double>(0, 1) *
+                   (best_position[i] - p.position[i]);
+    v = std::min(v, max_velocity);
+    v = std::max(v, min_velocity);
+
+    p.velocity[i] = v;
+  }
+}
+
+template <size_t Dimension, size_t ParticleNum>
+void PSO<Dimension, ParticleNum>::UpdateLocalBest() {
+  auto& p = swarm[idx];
+
+  if (unlikely(time == 0)) {
+    p.best_position = p.position;
+    p.best_fitness = p.fitness;
+    return;
+  }
+
+  if (p.fitness < p.best_fitness ^
+      !opt_minimize) {  // not when optimize to maximize
+    p.best_position = p.position;
+    p.best_fitness = p.fitness;
+  }
+}
+
+template <size_t Dimension, size_t ParticleNum>
+void PSO<Dimension, ParticleNum>::UpdateGlobalBest() {
+  if (unlikely(time == 0)) {
+    best_fitness = swarm[0].best_fitness;
+    best_position = swarm[0].best_position;
+  }
```

Comment:
As I said in MOptOptimizer, I don't like to see `time` is used here.
I think it would be better to remove `time` and initialize these values with some sentinels instead so that `UpdateLocalBest` and `UpdateGlobalBest` can be safely called any time in any order.

Reply:
fixed in b6207e50d21b625fb526c646059baf36fe5a43c4